src/dumbo/src/tcp/endpoint.rs

```diff
@@ -299,16 +303,36 @@ impl Endpoint {
     }
 }
 
+/// `parse_request` parses the request bytes and builds a `micro_http::Response` by the given callback function.
+pub fn parse_request(
+    request_bytes: &[u8],
+    callback: fn(Result<Request, RequestError>) -> Response,
+) -> Response {
+    let request = Request::try_from(request_bytes);
```

Comment:
```suggestion
    let request = Request::try_from(request_bytes)?;
```

Reply:
Hmm, it seems that this way we won't build the responses that were built before in the `Err(e)` match branch from `parse_request()` (or `callback()` now).
Are you thinking we should move that part somewhere else? It seems okay to build all the possible responses in the same place; maybe `callback()` should be somehow split in smaller functions, but I don't think we should treat those "error" cases in an upper level (as you mentioned I guess in a comment above).
Sorry in advance if I'm missing something, it's kind of the first time I'm looking at this code :sweat:.

Reply:
Sorry for the late reply, your proposal sounds good! From what it seems, there shouldn't be a lot of effort involved, at least from the implementation point of view; in tests, on the other hand, I think there will be several changes involved :-?. @moricho what do you think about this proposal? Does it make sense to you?

Reply:
@iulianbarbu @lauralt It makes sense to me. Thanks for the detailed explanation! Based on your proposal, I'm now refactoring that part. 
In that regard, could I ask you to give me some advice?  Do you think it's reasonable that we implement a function to build error response at `endpoint` level to treat `parse_request` error, aside from existing one to build mmds response (`build_response` func) at `mmds` crate ?

Reply:
@moricho , build response at the endpoint level makes sense for me, for responding to HTTP request parse errors.

I see it like this:

Dumbo assures the HTTP request is valid and if so it is passing it to MMDS. If the HTTP request is not valid, dumbo will emit a response corresponding to the `RequestError` encountered while parsing.

Before your PR, `RequestError` handling was done inside MMDS, where the request buffer was also parsed. Then, `mmds::build_response` was returning MMDS related errors (e.g `MmdsError::NotFound`) and micro-http request parsing error as well (e.g. `RequestError::InvalidUri`). Now, because we want to move the HTTP requests parsing in dumbo, we have to move `RequestError` handling (building and sending an HTTP response in case of parse error) in dumbo. Here, we can use a `build_response` function similar to `build_response` function from MMDS.

Let me know if this makes sense.

Reply:
@iulianbarbu Thanks for your advice! It makes sense for me. Based on this discussion, I'll do refactoring continuously.